### ISSUE TITLE
improper_sfs_exception: what():  SFS is not a probability distribution

### ISSUE THREAD

**Issue by grahamgower** [popgenmethods/smcpp#39]
Hi,
I've attempted to run SMC++ on a composite dataset containing a single diploid genome from shotgun sequencing (the distinguished individual) and SNP array data from 7 additional individuals (35 thousand loci). Here is an input file I created for this: [European.chr29.smc++.txt.gz](https://github.com/popgenmethods/smcpp/files/1298558/European.chr29.smc.txt.gz). Unfortunately, SMC++ is complaining that the SFS is not a probability distribution (see log below).

I added a few std::cout statements to see what was going on (see diff), but it didn't help me much. Any clues?

Thanks,
-Graham

```
diff --git a/src/conditioned_sfs.cpp b/src/conditioned_sfs.cpp
index 321ab6c..f3310d9 100644
--- a/src/conditioned_sfs.cpp
+++ b/src/conditioned_sfs.cpp
@@ -107,8 +107,14 @@ std::vector<Matrix<T> > incorporate_theta(const std::vector<Matrix<T> > &csfs, d
         assert(csfs[i](0, 0) == 0.);
         assert(csfs[i](2, n) == 0.);
         T tauh = csfs[i].sum();
-        if (toDouble(tauh) > 1.0 / theta)
+        if (toDouble(tauh) > 1.0 / theta) {
+            std::cout << csfs.size() << std::endl;
+            std::cout << csfs[i].size() << std::endl;
+            std::cout << i << std::endl << csfs[i].template cast<double>() << std::endl;
+            std::cout << tauh << std::endl;
+            std::cout << theta << std::endl;
             throw improper_sfs_exception();
+       }
         try
         {
             CHECK_NAN(tauh);
```

```
70223 smcpp.logging DEBUG SMC++ v1.9.5.dev0+g64d5243.d20170913
70224 smcpp.commands.command DEBUG ['/data/acad/apps/software/smcpp/20170701/bin/smc++', 'estimate', '-vv', '-o', '09_SMC/European', '--thinning', '0', '1.25e-8', '09_SMC/European.chr29.smc++.txt']
70224 smcpp.commands.command DEBUG Namespace(algorithm='L-BFGS-B', blocks=None, command='estimate', data=['09_SMC/European.chr29.smc++.txt'], em_iterations=20, ftol=0.0001, knots=10, length_cutoff=10000, mu=1.25e-08, no_filter=False, no_initialize=False, nonseg_cutoff=None, offset=0.0, outdir='09_SMC/European', polarization_error=0.5, r=None, regularization_degree=2, regularization_penalty=7, seed=0, spline='pchip', t1=None, tK=None, thinning=0, unfold=False, verbose=2, xtol=0.1)
70224 smcpp.analysis DEBUG Polarization error p=0.500000
70225 smcpp.analysis DEBUG regularization degree=2
70225 smcpp.analysis INFO Loading data...
70679 smcpp.estimation_tools WARNING File 09_SMC/European.chr29.smc++.txt doesn't appear to be in SMC++ format
70763 smcpp.analysis INFO 0.05 Gb of data
70764 smcpp.analysis DEBUG Contig(pid=('pop1',), fn='09_SMC/European.chr29.smc++.txt', n=array([14]), a=array([2]))
70765 smcpp.analysis INFO 1 population
71149 smcpp.analysis DEBUG Watterson estimator: 0.906175
71173 smcpp.analysis DEBUG Heterozygosity: 0.001547
71173 smcpp.analysis DEBUG Average heterozygosity (derived / total bases) by data set (* = dropped)
71174 smcpp.analysis DEBUG 09_SMC/European.chr29.smc++.txt:
71174 smcpp.analysis DEBUG               0       51505223       51505224  0.00154714
71174 smcpp.analysis INFO theta: 0.000100
71174 smcpp.analysis INFO rho: 0.000100
71174 smcpp.analysis DEBUG Ne: 9061.745526
71174 smcpp.analysis DEBUG setting t1=63850.012092
71174 smcpp.analysis DEBUG setting tK=223066317.373941
71175 smcpp.analysis DEBUG knots in coalescent scaling:
[      7.98125151      18.04672075      40.80614793      92.26838115     208.63165459     471.74521496    1066.68160339
    2411.91559962    5453.67693716   12331.52276951   27883.28967174   55766.57934349  111533.15868697]
71176 smcpp.analysis DEBUG Long term avg. effective population size: 7.735720
71176 smcpp.analysis DEBUG Creating inference manager...
71177 smcpp.analysis DEBUG Creating inference manager for (('pop1',), (14,), (2,))
71181 smcpp._smcpp:src/matrix_cache.cpp:62 DEBUG1 Acquired lock in init_cache()
71181 smcpp._smcpp:src/matrix_cache.cpp:74 DEBUG1 init_cache() successful
71196 smcpp._smcpp:src/inference_manager.cpp:236 DEBUG1 parallel filling targets
71287 smcpp._smcpp:src/inference_manager.cpp:252 DEBUG1 reducing targets
71319 smcpp._smcpp:src/inference_manager.cpp:53 DEBUG1 creating HMM
71780 smcpp.analysis INFO Running E-step
71782 smcpp._smcpp:src/inference_manager.cpp:108 DEBUG1 E step
71783 smcpp._smcpp:src/conditioned_sfs.cpp:88 DEBUG1 compute called
71783 smcpp._smcpp:src/conditioned_sfs.cpp:47 DEBUG1 compute above
71831 smcpp._smcpp:src/conditioned_sfs.cpp:15 DEBUG1 compute below
71832 smcpp._smcpp:src/conditioned_sfs.cpp:20 DEBUG1 tjj_double_integral below starts
71832 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
71846 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
71846 smcpp._smcpp:src/conditioned_sfs.cpp:24 DEBUG1 tjj_double_integral below finished
71846 smcpp._smcpp:src/conditioned_sfs.cpp:25 DEBUG1 matrix products below (M0)
71847 smcpp._smcpp:src/conditioned_sfs.cpp:27 DEBUG1 matrix products below (M1)
71847 smcpp._smcpp:src/conditioned_sfs.cpp:29 DEBUG1 filling csfs_below
71847 smcpp._smcpp:src/conditioned_sfs.cpp:37 DEBUG1 compute below finished
71847 smcpp._smcpp:src/conditioned_sfs.cpp:95 DEBUG1 compute finished
71847 smcpp._smcpp:src/inference_manager.cpp:407 DEBUG1 recompute B
71849 smcpp._smcpp:src/inference_manager.cpp:470 DEBUG1 recompute done
71849 smcpp._smcpp:src/transition.cpp:335 DEBUG1 computing transition
71872 smcpp._smcpp:src/transition.cpp:337 DEBUG1 done computing transition
71935 smcpp._smcpp:src/hmm.cpp:52 DEBUG1 forward algorithm (HMM #0)
71982 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 9%
72919 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 19%
72954 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 29%
72991 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 39%
73027 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 49%
73065 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 59%
73101 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 69%
73138 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 79%
73176 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 89%
73213 smcpp._smcpp:src/hmm.cpp:61 DEBUG1 hmm 0: 99%
73213 smcpp._smcpp:src/hmm.cpp:98 DEBUG1 backward algorithm (HMM #0)
73737 smcpp.analysis INFO E-step completed
73741 smcpp._smcpp:src/inference_manager.cpp:115 DEBUG1 InferenceManager::Q
75477 smcpp._smcpp:src/hmm.cpp:153 DEBUG1 HMM::Q
75477 smcpp._smcpp:src/inference_manager.cpp:125 DEBUG1 
q1:0 []
q2:-612225 []
q3:0 []

75478 smcpp._smcpp DEBUG im((('pop1',), (14,), (2,))).q1: 0.0
75478 smcpp._smcpp DEBUG im((('pop1',), (14,), (2,))).q2: -612225.1
75478 smcpp._smcpp DEBUG im((('pop1',), (14,), (2,))).q3: 0.0
75483 smcpp.analysis DEBUG reg: 0.0 []
75483 smcpp.analysis DEBUG Q:   -612225.1 []
75484 smcpp.analysis DEBUG Auto-assigning regularization penalty lambda=0.0612225
75485 smcpp.analysis DEBUG hidden states: [      0.               7.98125151      18.04672075      40.80614793      92.26838115     208.63165459     471.74521496
    1066.68160339    2411.91559962    5453.67693716   12331.52276951   27883.28967174   55766.57934349  111533.15868697
              inf]
75488 smcpp.analysis DEBUG knots in coalescent scaling:
[      7.98125151      18.04672075      40.80614793      92.26838115     208.63165459     471.74521496    1066.68160339
    2411.91559962    5453.67693716   12331.52276951   27883.28967174   55766.57934349  111533.15868697]
75488 smcpp.analysis DEBUG Long term avg. effective population size: 7.735720
75489 smcpp.analysis DEBUG Creating inference manager...
75494 smcpp.analysis DEBUG Creating inference manager for (('pop1',), (14,), (2,))
75495 smcpp._smcpp:src/matrix_cache.cpp:62 DEBUG1 Acquired lock in init_cache()
75495 smcpp._smcpp:src/matrix_cache.cpp:74 DEBUG1 init_cache() successful
75498 smcpp._smcpp:src/inference_manager.cpp:236 DEBUG1 parallel filling targets
75561 smcpp._smcpp:src/inference_manager.cpp:252 DEBUG1 reducing targets
75569 smcpp._smcpp:src/inference_manager.cpp:53 DEBUG1 creating HMM
75827 smcpp.optimize.plugins.progress_printer INFO Starting EM algorithm...
75827 smcpp.optimize.plugins.progress_printer INFO EM iteration 1 of 20...
75827 smcpp.analysis INFO Running E-step
75828 smcpp._smcpp:src/inference_manager.cpp:108 DEBUG1 E step
75828 smcpp._smcpp:src/conditioned_sfs.cpp:88 DEBUG1 compute called
75829 smcpp._smcpp:src/conditioned_sfs.cpp:47 DEBUG1 compute above
77629 smcpp._smcpp:src/conditioned_sfs.cpp:15 DEBUG1 compute below
77630 smcpp._smcpp:src/conditioned_sfs.cpp:20 DEBUG1 tjj_double_integral below starts
77630 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77630 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77630 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77630 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77630 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77631 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77631 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77631 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77632 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77632 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77632 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77632 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77633 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77633 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:304 DEBUG1 in tjj_double_integral_below
77633 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77634 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77634 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77635 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77635 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77636 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77637 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77638 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77638 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77638 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77639 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77640 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77640 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77640 smcpp._smcpp:src/piecewise_constant_rate_function.cpp:332 DEBUG1 exiting tjj_double_integral_below
77640 smcpp._smcpp:src/conditioned_sfs.cpp:24 DEBUG1 tjj_double_integral below finished
77641 smcpp._smcpp:src/conditioned_sfs.cpp:25 DEBUG1 matrix products below (M0)
77641 smcpp._smcpp:src/conditioned_sfs.cpp:27 DEBUG1 matrix products below (M1)
77642 smcpp._smcpp:src/conditioned_sfs.cpp:29 DEBUG1 filling csfs_below
77642 smcpp._smcpp:src/conditioned_sfs.cpp:37 DEBUG1 compute below finished
77642 smcpp._smcpp:src/conditioned_sfs.cpp:95 DEBUG1 compute finished
14
45
9
         0    12.7436    5.17706    2.76109    1.62707    1.00127   0.625793   0.390104   0.238938   0.141593  0.0796457  0.0413743  0.0189632 0.00700178 0.00162541
   242.903    451.401    627.702    771.859      883.9    963.845    1011.71     1027.5    1011.22    962.893    882.512    770.085    625.616     449.11     240.57
         0          0          0          0          0          0          0          0          0          0          0          0          0          0          0
10947.7
0.0001
terminate called after throwing an instance of 'improper_sfs_exception'
  what():  SFS is not a probability distribution
stack trace:
  /lib64/libc.so.6() [0x3fafc326a0]
  /lib64/libc.so.6 : gsignal()+0x35
  /lib64/libc.so.6 : abort()+0x175
  /apps/software/GCCcore/5.4.0/lib64/libstdc++.so.6 : __gnu_cxx::__verbose_terminate_handler()+0x15d
  /apps/software/GCCcore/5.4.0/lib64/libstdc++.so.6 : ()+0x96786
  /apps/software/GCCcore/5.4.0/lib64/libstdc++.so.6 : ()+0x967d1
  /apps/software/GCCcore/5.4.0/lib64/libstdc++.so.6 : ()+0x969e8
  /data/acad/apps/software/smcpp/20170701/lib/python3.6/site-packages/smcpp-1.9.5.dev0+g64d5243.d20170913-py3.6-linux-x86_64.egg/smcpp/_smcpp.cpython-36m-x86_64-linux-gnu.so : std::vector<Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1, 0, -1, 1> >, -1, -1, 0, -1, -1>, std::allocator<Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1, 0, -1, 1> >, -1, -1, 0, -1, -1> > > incorporate_theta<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1, 0, -1, 1> > >(std::vector<Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1, 0, -1, 1> >, -1, -1, 0, -1, -1>, std::allocator<Eigen::Matrix<Eigen::AutoDiffScalar<Eigen::Matrix<double, -1, 1, 0, -1, 1> >, -1, -1, 0, -1, -1> > > const&, double)+0x1a5f
  /data/acad/apps/software/smcpp/20170701/lib/python3.6/site-packages/smcpp-1.9.5.dev0+g64d5243.d20170913-py3.6-linux-x86_64.egg/smcpp/_smcpp.cpython-36m-x86_64-linux-gnu.so : NPopInferenceManager<1ul>::recompute_emission_probs()+0x10e
  /data/acad/apps/software/smcpp/20170701/lib/python3.6/site-packages/smcpp-1.9.5.dev0+g64d5243.d20170913-py3.6-linux-x86_64.egg/smcpp/_smcpp.cpython-36m-x86_64-linux-gnu.so : InferenceManager::do_dirty_work()+0x28
  /data/acad/apps/software/smcpp/20170701/lib/python3.6/site-packages/smcpp-1.9.5.dev0+g64d5243.d20170913-py3.6-linux-x86_64.egg/smcpp/_smcpp.cpython-36m-x86_64-linux-gnu.so : InferenceManager::Estep(bool)+0xa0
  /data/acad/apps/software/smcpp/20170701/lib/python3.6/site-packages/smcpp-1.9.5.dev0+g64d5243.d20170913-py3.6-linux-x86_64.egg/smcpp/_smcpp.cpython-36m-x86_64-linux-gnu.so : ()+0x54a19
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : PyCFunction_Call()+0xf9
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyEval_EvalFrameDefault()+0x726e
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140070
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140db6
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyEval_EvalFrameDefault()+0x4c26
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140a1c
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : PyEval_EvalCodeEx()+0x2f
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0xafd35
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : PyObject_Call()+0x6a
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyEval_EvalFrameDefault()+0x470e
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140070
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140db6
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyEval_EvalFrameDefault()+0x4c26
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140070
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140db6
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyEval_EvalFrameDefault()+0x4c26
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x140070
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyFunction_FastCallDict()+0x2d8
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyObject_FastCallDict()+0x166
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : _PyObject_Call_Prepend()+0xcd
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : PyObject_Call()+0x6a
  /data/acad/apps/software/Python/3.6.2-foss-2016b/lib/libpython3.6m.so.1.0 : ()+0x184cf2
  /lib64/libpthread.so.0() [0x3fb00079d1]
  /lib64/libc.so.6 : clone()+0x6d
```

**Comment by grahamgower** [comment]
One additional thing which may be relevant is this population is highly inbred, as can be seen from the Ne of 7.73. Actually, the census population size was as low as seven ~80 years ago.

**Comment by terhorst** [comment]
It's a bug in SMC++, the calculation of Watterson's estimator is wrong. It'll be fixed in the next release.

**Comment by grahamgower** [comment]
I played around with this today and discovered (1) Watterson's estimator needs at least 2 haplotypes, and (2) the sample size was calculated incorrectly. With the patch below, SMC++ runs successfully and the output roughly matches that of MSMC.

```
diff --git a/smcpp/estimation_tools.py b/smcpp/estimation_tools.py
index 9e95ef7..172d269 100644
--- a/smcpp/estimation_tools.py
+++ b/smcpp/estimation_tools.py
@@ -182,10 +182,10 @@ def watterson_estimator(contigs):
         num = denom = 0
         for S, sample_sizes, spans in map(_watterson_helper, contigs):
             num += S
-            non_missing = sample_sizes > 0
+            non_missing = sample_sizes > 1
             ss = sample_sizes[non_missing]
             sp = spans[non_missing]
-            denom += (sp * (np.log(ss) + 0.5 / ss + 0.57721)).sum()
+            denom += (sp * (np.log(ss) + 0.5 / ss + np.euler_gamma)).sum()
     return num / denom
 
 
@@ -197,8 +197,8 @@ def _watterson_helper(contig):
     seg = (np.any(c.data[:, 1::3] >= 1, axis=1) |
            np.any(c.data[:, 2::3] >  0, axis=1))
     S = spans[seg].sum()
-    sample_sizes = (c.data[:, 3::3].sum(axis=1) +
-                    np.maximum(0, c.data[:, 1::3]).sum(axis=1))
+    sample_sizes = (c.data[:, 3::2].sum(axis=1) +
+                    list(map(lambda a: 0 if a==-1 else 2, c.data[:, 1])))
     return (S, sample_sizes, spans)
 
 def _load_data_helper(fn):
```

**Comment by terhorst** [comment]
I fixed this in a different way as part of a larger refactoring effort that is now released as v1.10.0. Thanks for submitting the patch.
